go-controller/pkg/ovn/egressfirewall.go

```diff
@@ -277,23 +290,30 @@ func (oc *Controller) updateEgressFirewall(oldEgressFirewall, newEgressFirewall
 
 func (oc *Controller) deleteEgressFirewall(egressFirewall *egressfirewallapi.EgressFirewall) error {
 	klog.Infof("Deleting egress Firewall %s in namespace %s", egressFirewall.Name, egressFirewall.Namespace)
-	deleteDNS := false
+	//deleteDNS := false
 
 	nsInfo := oc.getNamespaceLocked(egressFirewall.Namespace)
 	if nsInfo != nil {
 		// clear it so an error does not prevent future egressFirewalls
 		for _, rule := range nsInfo.egressFirewall.egressRules {
 			if len(rule.to.dnsName) > 0 {
-				deleteDNS = true
-				break
+				oc.egressfirewallDNSMutex.Lock()
+				delete(oc.egressfirewallDNSInfo[rule.to.dnsName].Namespaces, egressFirewall.Namespace)
+
+				if len(oc.egressfirewallDNSInfo[rule.to.dnsName].Namespaces) == 0 {
+					// there is no namespace using the egressfirewall rule so it is safe to delete the addressSet
+					err := oc.egressfirewallDNSInfo[rule.to.dnsName].as.Destroy()
+					if err != nil {
```

Comment:
you are not unlocking here.
You can use defer in embedded blocks (and avoid forgetting) using anonymous functions
```
func () {
    lock()
    defer unlock()
}()
```

Reply:
fixed

Reply:
@fedepaol I am resolving this conversation as I think I took care of it, if I was not correct could you mark this as unsolved?

Reply:
looks good, thanks